src/java.base/share/classes/jdk/internal/foreign/abi/fallback/FFIType.java

```diff
@@ -0,0 +1,144 @@
+
+    private static MemorySegment make(List<MemoryLayout> elements, FFIABI abi, SegmentScope scope) {
+        MemorySegment elementsSeg = MemorySegment.allocateNative((elements.size() + 1) * ADDRESS.byteSize(), scope);
+        int i = 0;
+        for (; i < elements.size(); i++) {
+            MemoryLayout elementLayout = elements.get(i);
+            MemorySegment elementType = toFFIType(elementLayout, abi, scope);
+            elementsSeg.setAtIndex(ADDRESS, i, elementType);
+        }
+        // elements array is null-terminated
+        elementsSeg.setAtIndex(ADDRESS, i, MemorySegment.NULL);
+
+        MemorySegment ffiType = MemorySegment.allocateNative(LAYOUT, scope);
+        VH_TYPE.set(ffiType, LibFallback.STRUCT_TAG);
+        VH_ELEMENTS.set(ffiType, elementsSeg);
+
+        return ffiType;
+    }
+
+    private static final Map<Class<?>, MemorySegment> CARRIER_TO_TYPE = Map.of(
+        boolean.class, LibFallback.UINT8_TYPE,
+        byte.class, LibFallback.SINT8_TYPE,
+        short.class, LibFallback.SINT16_TYPE,
+        char.class, LibFallback.UINT16_TYPE,
+        int.class, LibFallback.SINT32_TYPE,
+        long.class, LibFallback.SINT64_TYPE,
+        float.class, LibFallback.FLOAT_TYPE,
+        double.class, LibFallback.DOUBLE_TYPE,
+        MemorySegment.class, LibFallback.POINTER_TYPE
+    );
+
+    static MemorySegment toFFIType(MemoryLayout layout, FFIABI abi, SegmentScope scope) {
+        if (layout instanceof GroupLayout grpl) {
+            if (grpl instanceof StructLayout) {
+                // libffi doesn't want our padding
+                List<MemoryLayout> filteredLayouts = grpl.memberLayouts().stream()
+                        .filter(Predicate.not(PaddingLayout.class::isInstance))
+                        .toList();
+                MemorySegment structType = make(filteredLayouts, abi, scope);
+                verifyStructType(grpl, filteredLayouts, structType, abi);
+                return structType;
+            }
+            assert grpl instanceof UnionLayout;
+            throw new UnsupportedOperationException("No unions (TODO)");
+        } else if (layout instanceof SequenceLayout sl) {
+            List<MemoryLayout> elements = Collections.nCopies(Math.toIntExact(sl.elementCount()), sl.elementLayout());
+            return make(elements, abi, scope);
+        }
+        return Objects.requireNonNull(CARRIER_TO_TYPE.get(((ValueLayout) layout).carrier()));
+    }
+
+    // verify layout against what libffi sets
+    private static void verifyStructType(GroupLayout grpl, List<MemoryLayout> filteredLayouts, MemorySegment structType,
+                                         FFIABI abi) {
+        try (Arena verifyArena = Arena.openConfined()) {
+            MemorySegment offsetsOut = verifyArena.allocate(SIZE_T.byteSize() * filteredLayouts.size());
+            LibFallback.getStructOffsets(structType, offsetsOut, abi);
```

Comment:
Maybe I'm being paranoid, but libFFI has its own way to determine a struct layout given the struct fields, based on what the ABI types are. But the Panama layout API is more expressive than that, and it allows, theoretically, to pass arguments that have stricter or larger alignment constraints - which can end up in more or less padding being inserted (from clients), and in different struct layout sizes.

So, my question is: should we double check that what we get back from libFFI is actually compatible with what we have in the group layout before proceeding?

Reply:
Yes, that is exactly what this code (`verifyStructType`) is trying to do.

Note that we also already reject layouts that do not have natural alignment [1](https://github.com/openjdk/panama-foreign/blob/foreign-memaccess%2Babi/src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java#L111) Though, we might want to relax that constraint in the future.

Reply:
More specifically, if a struct has extra padding, then I wonder if libFFI could expect a struct size that is smaller than the size of the memory segment passed in by a Panama client (because that segment also has the extra padding). In our usual JDK-driven classification, we would take the struct, extract its field components and place them on registers and/or stack slots. But with libFFI, we just have to provide the struct segment in a buffer, which then libFFI will act upon - if there is a size mismatch there, I think it might have dire consequences?

Reply:
Anyway - I appreciate that the code verifies for mismatches. Good job. I was worried about the analysis being too "field-oriented", but you also look at offsets, so any extra padding should result in issues. Well, except maybe a trailing padding field?